gtfs-realtime/spec/en/trip-modifications.md

```diff
@@ -0,0 +1,64 @@
+# Trip Modifications
+
+A `TripModifications` message identifies a list of similar `trip_ids` from the (CSV) GTFS sharing the same stop pattern, which are all affected by particular modifications, such as a detour.
+
+## SLO: Service-level objective
+
+The frequency of data updates is expected to be approximately hourly (~24 times/day). Ingestion time may depend on the total number of affected trips. Consumers are expected to ingest a single TripModification within 5 minutes, and a feed with hundreds of detours within 20 minutes.
+
+## TripModifications
+
+The `TripModifications` is in effect on all of the listed service\_dates, until it is removed from the feed. On any given service date, a trip MUST NOT be assigned to more than one `TripModifications` object.
+
+There MAY be multiple `TripModifications` for a given stop pattern. It may be desirable to split the trips into multiple modifications e.g. if the `propagated_modification_delay` changes significantly, over the course of the detour.
+
+The trips created through GTFS-TripModifications modify and replace each specified `trip_id`, and don't create a copy or additional run. Modifications are applied on the schedule information, like if a static GTFS (CSV) was modified. 
+
+The scheduled stop times of each replacement trip are created from those of the affected trip, by performing the changes listed in modifications. `stop_sequence` for all stop times are replaced by a new value of 1 to n, starting with 1 on the first stop_time and increasing by 1 for each stop in the trip. The trip ID is also replaced by a replacement trip ID (`modifications_id` + `_` + `trip_id`). A `TripUpdate` message must be provided to publish real-time arrival/departure times for the replacement trip.
+
+
+## Linkage to TripUpdates
```

Comment:
I had a question about the linkage to TripUpdates.  I recognize there has been some back-and-forth on this in the comments, based on feedback from IBI Group and others, and I'm reflecting on the behavior of what happens for consumers who haven't updated yet to support trip modifications.

Apologies if I missed this up-thread, but I'm guessing you considered (and discarded) the option to just supply a TripUpdate with the existing `trip_id` that includes updates for:
* existing stops (no real change here)
* new stops in the trip (with possibly a new StopTimeUpdate. ScheduleRelationship enum value)
* skipped stops (either not mentioned at all or explicitly marking them SKIPPED)

When I think about consumers who might be slow to support trip modifications directly, having the separate trip update with a modified trip id would require some level of modification processing.  On the other hand, if the trip id was the same, these consumers might be able to more gracefully fallback to still apply stop time updates to their existing concept of the trip and hide the skipped stops (assuming they support SKIPPED).

I recognize `StopTimeUpdate.stop_sequence` indexing gets a bit murky here.

Reply:
The current text allows using the existing `trip_id` to affect existing stop and put SKIPPED on stops that are SKIPPED by the detour. That's what Swiftly is doing (and was doing before TripModifications too). 

On addressing new stops : I'm a bit worried that this would allow strange behaviour, validators will point that out as an error, etc. And yes, it wouldn't be possible to target stops with `stop_sequence`. You would also still need to ingest the new Stops. I feel having a light or partial implementation of TripModification would make more sense to me than try to allow everything in TripUpdates. 

One alternative is to use the experimental `stop_time_properties.assigned_stop_id` to assign new stops in real time for consumer that supports this and not TripModifications.

Reply:
To be clear, I think I'm mostly fine with the high-level concept of a first-class "Trip Modification" itself (aka I'm not proposing to cram all this behavior into TripUpdates only).  But I'm keen to understand what TripUpdates would look like in a feed with modifications.

Swiftly has already mentioned that they are going down the path of publishing separate GTFS-Realtime feeds based on whether a consumer can handle trip modifications or not.  I know that's not the first instance of this happening with GTFS data in general, but it's still not a great precedent to set.  I'd rather just have one feed where a consumer who doesn't support TripModifications fully could still get some value out of the feed.  For example, we've already discussed why some consumers might not be able to support detours more generally, but most can realistically handle the skipped stops and stop-time updates for stops that haven't changed along the route.

But maybe this is all just hand-waving around my specific concern: I don't really like the "new trip id" = `modifications_id` + `_` + `trip_id` convention.  I'd rather put `modification_id` as a first-class field in `TripDescriptor` if you need that linkage.  But why is this modification id needed here at all?  If a TripUpdate specifies a trip id for a trip that has an active modification, then that linkage is already specified by the TripModification itself, yes?  Is there are world where a feed would include a TripUpdate for the non-modified trip and for the modified trip as the same time?

Reply:
> Is there are world where a feed would include a TripUpdate for the non-modified trip and for the modified trip as the same time?

Swiftly does include TripUpdates for both the non-modified trip and the modified trip (when serving a request that wants modified trips). This is so there is no gap in real-time information between the time that we know about a modification, and the time that the consumer is able to ingest and apply it. To quite @mads14's [comment](https://github.com/google/transit/pull/403#issuecomment-1758406979):

> Meanwhile, we do still include all non-modified trip ids in our feed as well with the understanding that the consumer will replace those with the modified one when they read it, but given that it will take consumers some time (5 min or more) to read the trip modification details we didn't want to omit predictions for the normal trip id in the meantime.

Reply:
Coming back to this point, I think the introduction of ModifiedTripSelector is an improvement over the id concatenation approach.  To the larger point, would it be worthwhile to add language below around the expectations for both producers and consumers for specifying updates for the original unmodified trip?  If we are arguing for backwards compatibility for existing clients, then I'd argue the unmodified trip should continue to be specified for the duration of the detour.  Thoughts?

Reply:
Having the unmodified and modified trip in the trip updates could cause consumers that support predictions from trips with unidentified trip_ids, such as TransSee, to generate predictions for both if they aren't modified to handle it.

Reply:
@bdferris-v2 I didn't see that update from yesterday, but I agree that producer SHOULD provide both. I can update the language. 

@doconnoronca Did you review the changes since that original comment was posted? Specifically https://github.com/google/transit/pull/403/files#diff-abc780f2e1f0fd946abf169b0bb5d5ed06c33c695a028c7119043c0584eb7759R23

Reply:
@gcamp Wouldn't that only work for consumers that have been updated to handle trip modifications?

Reply:
No, since there would be two trip updates, one with a "regular" `TripDescriptor` exactly like there is today (most of the time, with just a `trip_id`) and a second one with a `TripDescriptor` empty except for the `ModifiedTripSelector`. 

Clients that know about trip modification would use the update that has a `ModifiedTripSelector`. That update should be ignored by clients that don't know about trip modifications since the `TripDescriptor` would be empty otherwise. 

Clients that don't know about trip modifications would just use the update with "regular" `TripDescriptor`.

Reply:
I see. Without a route_id the prediction couldn't be used without processing the ModifiedTripSelector.